traits/has_traits.py

```diff
@@ -2658,20 +2658,27 @@ def on_trait_change(
                 if wrapper.equals(handler):
                     break
             else:
-                listener = ListenerParser(name).listener
-                lnw = ListenerNotifyWrapper(
-                    handler, self, name, listener, target
-                )
-                listeners.append(lnw)
-                listener.trait_set(
+                # The listener notify wrapper needs a reference to the
+                # listener, and the listener needs a (weak) reference to the
+                # wrapper. We first construct the wrapper with a listener of
+                # `None`, then construct the listener with its reference to the
+                # wrapper, then we replace the `None` listener with the correct
+                # one.
+                lnw = ListenerNotifyWrapper(handler, self, name, None, target)
+                listener = ListenerParser(
+                    name,
                     handler=ListenerHandler(handler),
                     wrapped_handler_ref=weakref.ref(lnw),
-                    type=lnw.type,
                     dispatch=dispatch,
+                ).listener
+                listener.trait_set(
+                    type=lnw.type,
                     priority=priority,
                     deferred=deferred,
```

Comment:
Why not make the same changes to these three traits as well? Is that planned in a subsequent PR? `type` and `deferred` don't seem to have any static trait change handlers hooked up to them but `priority` does.

Reply:
Subsequent PR. For `type` and `deferred`, it's messy, as you'll see when the next PR lands. :-) `priority` is _mostly_ the same as in this PR, except that it's handled differently at `ListenerGroup` level.